### ISSUE TITLE
Usage in multi-module project and/or usage without decorators

### ISSUE THREAD

**Issue by tuukkamustonen** [coleifer/huey#196]
I have two web applications A and B, that send messages to a remote server. There is a third library C that both A and B depend on - C contains the actual message sending code. It works.

Now, for technical reasons, I want messages to be sent async. So... task queue for the rescue!

So, I would like to add app D as a consumer application. I would keep message sending code in C, and D would simply read configs and construct `RedisHuey` instance (as would A and B). However, if I am to decorate my task as `@huey.task`, what should I import?

Could I solve this by:

1. Can I adjust `RedisHuey` settings (e.g. Redis URL) after tasks are loaded? If I put just `huey = RedisHuey()` to C and then import it into A/B/D and reconfigure, would that work?

2. `rq` allows to enqueue tasks like:

        q.enqueue(some_task, 'parameter')

    I see huey has similar more low-level `Huey.enqueue(task)` API that it internally calls. Could or should I, instead of using decorators, do something like:

        huey.enqueue(create_task(QueueTask, my_task_function, ...))

    Then I wouldn't have to decorate my task and import huey instance into C.

Hopefully I'm not over-thinking it :|

**Comment by tuukkamustonen** [comment]
Btw. I tested option #1 it so that in `tasks.py`:

    huey = RedisHuey()

    @huey.task()
    def foo():
        ...

In `main.py`:

    from tasks import huey
    
    def build_app(conf_dir):
        # load configs and whatever

        huey.__init__(...)

It seems to work but is ugly as... Peewee has `Proxy`, could I re-use it here?

**Comment by tuukkamustonen** [comment]
Peewee's `Proxy` works just fine, but I'd like to suggest alternative API that supports calling class instances:

`app.py`:

```python
from x import MessageClient
from huey import RedisHuey
from messenger import Messenger

def build_app():

    client = MessageClient()
    messenger = Messenger(client)

    huey = RedisHuey()
    huey.add_instance_task(messenger.send_message)  # register task handler into class instance

    return huey
```

`messenger.py`:

```python
from huey import task

class Messenger:

    def __init__(self, client):
        self.client = client

    @task()  # would not register task, but wrap it either queueing to Redis or passthrough if always_eager is True
    def send_message(self, message):  # self parameter would NOT be pickled/unpickled
        self.client.send(...)
```

Run `huey_consumer app.huey` as usual.

This way I wouldn't have to play with proxies, I could use OO and different configurations in producer and consumer apps, and I wouldn't have to worry about import order.

What do you think @coleifer ?

**Comment by coleifer** [comment]
I'm going to pass. This sounds like a problem that arose out of your architecture, patterns, etc, rather than something particular to Huey. Why not just use a config file and configure huey using values taken from that? It can be that simple!

**Comment by tuukkamustonen** [comment]
Sorry, I was busy in other experiments.

Yeah, it's about architecture. Rather than declaring stuff on module level, I tend to build applications in factories as suggested by Flask (http://flask.pocoo.org/docs/0.12/patterns/appfactories/) or DropWizard (http://www.dropwizard.io/1.0.5/docs/getting-started.html). I also like to inject dependencies when building the app and services, rather than accessing dependencies through globals/threadlocals (e.g. Flask's current_app.something) or by declaring instances at module level (or by not doing classes at all). This makes testing (mocking) so much easier and avoids a few problems with import order etc.

In my case, I would like to intialize huey the same way I initialize a Flask app (=app factory), so it could re-use the existing code.

Yeah, there are workarounds to bolt huey to such architecture. To load config:

1. The one I referred to earlier: using peewee's `Proxy`.

2. What you suggested: just loading the config and instantiate huey within the same module:

Whichever the approach, the problem is this:

`config.py`:

```python
from huey import RedisHuey

config = ...  # load config based on os.environ['CONFIG_PATH'] or something
huey = RedisHuey(config)
```

`app.py`:

```python
def build_app():
    ...

    client = MessageClient(config['url'])
    notifier = Notifier(client)

    from config import huey  # to initialize
    import tasks  # to run the task decorators
```

`tasks.py`:

```python
from huey import huey

@huey.task()
def send_msg(message):
    ...  # the problem is here, how can I get hold of `notifier` that was built in build_app()?
```

Workarounds to access `notifier`:

1. Attach instances to huey instance (or it could be anything globally accessible), for example in `app.py`:

    ```python
    def build_app():
        ...
    
        huey.notifier = notifier
    ```

    `tasks.py`:
    
    ```python
    @huey.task()
    def send_msg(message):
        huey.notifier.send(message)
    ```

2. Re-construct dependencies within the function like:

    ```python
    @huey.task()
    def send_msg(message):
        ...  # load configs based on os.environ['CONFIG_PATH'] etc.
        client = MessageClient(config['url'])
        notifier = Notifier(client)
        huey.notifier.send(message)
    ```

    Now this would be super clumsy.

3. Don't use classes and just do functions... uh...

So... there are options, but it gets ugly. And testing gets difficult. Maybe I'm missing something here...

---

In any case, I tried decorating the instance method on the fly:

```python
def build_app():
    ...

    huey = RedisHuey('messages')
    notifier.send = huey.task()(notifier.send)

    ...
```

In `service.py`:

```python
class Notifier:
    def __init__(self, client):
        self.client = client

    def send(self, message):
        self.client.send...
```

So rather than decorating a function, I'm decorating an instance method here.

The problem with this is that huey saves attributes to the decorated function (https://github.com/coleifer/huey/blob/master/huey/api.py#L107-L108), and that's not allowed for instance method. So, I commented out those lines for a test, and it seems to work (it also avoids pickling `self`).

So... could we store the function attributes to a separate registry (e.g. dict) in `Huey`? Would that change break/prevent something? It would open up this usage pattern (OO, app factories, IOC, whatever is the right term, I believe it's pretty common).

(Finally, if user wants to pickle the whole class instance, which sounds risky performance & security wise, he can just decorate a method as usual so no change of behavior here:)

```python
class Notifier:
    ...

    @heuy.task()
    def send(self, message):
        ...
```
